Support for STM32L0 MCUs and Dragino LGT92 device

ofauchon opened the pull request:
This PR provides support for : 

- STM32L0 microcontroller series. 
- Dragino LGT92 device (which is based on stm32l072czt6  )

Working:  

 - LPUART and UART
 - GPIOs
 - SPI
 - Timers

Not working yet: 

- i2c
- power saving (LP modes)

ofauchon committed:
Add support for stm32l0 familly, and Dragino LGT92 Board

ofauchon committed:
- Dead code removed
- lgt92.json now inherits stm32l0x2.json

ofauchon committed:
removed lgt92 build tag in stm32l0x2.json

aykevl commented on src/machine/board_lgt92.go:
```diff
@@ -0,0 +1,107 @@
+)
+
+const (
+	LED1 = PA12
```
Can you also add this?

```go
	LED = LED1
```

Many examples use `machine.LED` for the first (or only) LED available.

ofauchon replied:
Done

aykevl commented on src/machine/board_lgt92.go:
```diff
@@ -0,0 +1,107 @@
+func init() {
+
+	// Enable GPIOs
+	stm32.RCC.IOPENR.SetBits(stm32.RCC_IOPENR_IOPAEN)
+	stm32.RCC.IOPENR.SetBits(stm32.RCC_IOPENR_IOPBEN)
+	stm32.RCC.IOPENR.SetBits(stm32.RCC_IOPENR_IOPCEN)
+
+	// Enable NVM (eeprom)
+	stm32.RCC.AHBENR.SetBits(stm32.RCC_AHBENR_MIFEN)
+
+	// Enable Leds GPIO output
+	LED_RED.Configure(PinConfig{Mode: PinOutput})
+	LED_GREEN.Configure(PinConfig{Mode: PinOutput})
+	LED_BLUE.Configure(PinConfig{Mode: PinOutput})
+
+	LED_RED.Low()
+	LED_GREEN.Low()
+	LED_BLUE.Low()
```
Why are you doing this in the board file?

- Configuring outputs needs to be done by the program, unless it's really necessary they should be configured by a program that wants to use these LEDs.
- Is there any support for EEPROM? If not, I would suggest leaving it disabled (enabling unused peripherals may cost power).
- If you need to enable the GPIO peripheral (`stm32.RCC.IOPENR.SetBits(stm32.RCC_IOPENR_IOPAEN)`), this should be done in `Pin.Configure`. It looks like this is already done, so these lines are unnecessary.

aykevl commented on src/machine/machine_stm32l0.go:
```diff
@@ -0,0 +1,193 @@
+		//	case 5:
+		//		return stm32.GPIOF
+		//	case 6:
+		//		return stm32.GPIOG
```
If `GPIOF` etc. don't exist, you don't need to include them here. If you want to explain why you left them out, you could add a comment saying "GPIOF etc. do not exist on this chip".

aykevl commented on src/machine/machine_stm32l0.go:
```diff
@@ -0,0 +1,193 @@
+		//	case unsafe.Pointer(stm32.TIM8): // TIM8 clock enable
+		//		stm32.RCC.APB2ENR.SetBits(stm32.RCC_APB2ENR_TIM8EN)
+		//	case unsafe.Pointer(stm32.TIM1): // TIM1 clock enable
+		//		stm32.RCC.APB2ENR.SetBits(stm32.RCC_APB)
```
This is again code that should be removed: TIM1 and TIM8 do not seem to exist on this chip.

aykevl commented on src/machine/machine_stm32l0x2.go:
```diff
@@ -0,0 +1,108 @@
+
+	// These are based on APB2 clock frquency (84MHz on the discovery board)
+	// TODO: also include the MCU/APB clock setting in the equation
+	switch true {
```
You can leave out the `true` here, a switch can also work as a replacement for a long `if` / `else if` / `else` chain.

```suggestion
	switch {
```

aykevl commented on src/machine/machine_stm32l0x2.go:
```diff
@@ -0,0 +1,108 @@
+		conf = stm32.SPI_PCLK_2
+	default:
+		// None of the specific baudrates were selected; choose the lowest speed
+		conf = stm32.SPI_PCLK_256
```
Ideally we'd default to 4MHz, but if that is not possible you could use a value just below that.
You can add such a check at the top by checking for 0:

```go
if config.Frequency == 0 {
    config.Frequency = 4e6
}
```

aykevl commented on src/machine/machine_stm32l0x2.go:
```diff
@@ -0,0 +1,108 @@
+	// These are based on APB2 clock frquency (84MHz on the discovery board)
+	// TODO: also include the MCU/APB clock setting in the equation
+	switch true {
+	case localFrequency < 328125:
+		conf = stm32.SPI_PCLK_256
```
It would be useful to make this calculation explicit. For example:

```go
const apb2Frequency = 84e6

// ...

switch {
case localFrequency >= apb2Frequency / 2:
    conf = stm32.SPI_PCLK_2
case localFrequency >= apb2Frequency / 4:
    conf = stm32.SPI_PCLK_4
// ...
case localFrequency >= apb2Frequency / 128:
    conf = stm32.SPI_PCLK_128
default:
    conf = stm32.SPI_PCLK_256
}
```

This way the calculation is very explicit. This kind of calculation is always done at compile time, so you don't need to worry about any overhead it might introduce.
You can take a look here as an example: https://github.com/tinygo-org/tinygo/pull/1446

aykevl commented on targets/stm32l0x2.json:
```diff
@@ -0,0 +1,22 @@
+        "src/device/stm32/stm32l0x2.s"
+    ],
+    "flash-method": "openocd",
+    "openocd-interface": "stlink-v2",
```
```suggestion
```

These two lines are specific to the board, so shouldn't be in the chip specification.

aykevl commented on targets/stm32l0x2.json:
```diff
@@ -0,0 +1,22 @@
+        "--target=armv6m-none-eabi",
+        "-Qunused-arguments"
+    ],
+    "linkerscript": "targets/stm32l072czt6.ld",
```
```suggestion
```

Because this linker script is for a more specific chip (with a given amount of memory), it shouldn't be included in the more generic chip family specification.

aykevl reviewed (commented):
Thank you, looks good to me! I have a number of code cleanliness issues but nothing major.

aykevl commented:
And before I forget: can you also add a smoke test to the Makefile? You can see a list of commands there already to run the smoke tests.

ofauchon committed:
Add smoketest for lgt92
Add default LED declaration so example compiles
Remove useless hardware init and comments
Move all machine_stm32l0x2 code back to machine_stm32l0 as it'll probably work for stm32l0x1
Move Flash-method,linkerscript... from chip target to board target json

ofauchon committed:
Moved runtime_stm32l0x2 in runtime_stm32l0

ofauchon committed:
Add some registers const to avoid the use of an extra stm32l0x2-bitfields.go file.

ofauchon committed:
Remove binary integer prefix 0b, for Golang<1.13 backward compatibility

ofauchon committed:
Remove more binary integer prefix 0b, for Golang<1.13 backward compatibility

ofauchon commented:
Hi @aykevl,

Thanks for your detailed review, and all the suggestions. 
I think I made all the proposed changes. 

Please tell me if you have other improvements.

Olivier

aykevl reviewed (approved):
Thank you for the fixes! Looks good to me now.

deadprogram commented:
I will squash/merge this PR now. Thank you very much for the contribution @ofauchon
